### PR TITLE
.github/workflows: add user documentation workflow

### PR DESCRIPTION
**Pull request by seilagamo**
Add a workflow that creates the user documentation automatically using [mdBook](https://rust-lang.github.io/mdBook/) to build the static site.

### CODE REVIEW

File: book.toml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[book]
```
Comment (jroimartin) [inline book.toml:3]: Add copyright notice.

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+book
```
Comment (jroimartin) [inline .gitignore]:
```suggestion
/book/
```

File: .github/workflows/doc.yaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+name: Documentation
+on:
+  push:
+    branches:
+    - mdBook
+permissions:
+  contents: read
+  pages: write
+  id-token: write
```
Comment (jroimartin) [inline .github/workflows/doc.yaml]: I would split the `build` job in 2 jobs: `build` and `deploy`. So, only `deploy` has the `pages: write` and `id-token: write` permissions. I would also add a comment explaining why these permissions are needed. Check file comment for more details.

File: .github/workflows/doc.yaml [hunk 4]
```diff
@@ -0,0 +1,37 @@
+# Copyright 2024 Adevinta
+
+name: Documentation
+on:
+  push:
+    branches:
+    - mdBook
```
Comment (jroimartin) [inline .github/workflows/doc.yaml]: Should we use `main` here? Check file comment for more details.

```suggestion
      - main
```

File: .github/workflows/doc.yaml
Comment (jroimartin) [inline .github/workflows/doc.yaml:1]: I would rewrite the workflow to something like this. Check the other comments for more details.

```yaml
# Copyright 2024 Adevinta

name: Deploy Lava documentation to Pages

on:
  push:
    branches:
      - main

permissions:
  contents: read

# Allow only one concurrent deployment, skipping runs queued between
# the run in-progress and latest queued. However, do not cancel
# in-progress runs as we want to allow these production deployments to
# complete.
concurrency:
  group: "pages"
  cancel-in-progress: false

jobs:
  build:
    name: Generate documentation
    runs-on: ubuntu-latest
    steps:
      - name: Checkout repository
        uses: actions/checkout@v4
      - name: Build documentation
        run: ./build.bash
      - name: Setup Pages
        uses: actions/configure-pages@v4
      - name: Upload artifact
        uses: actions/upload-pages-artifact@v3
        with:
          path: 'book'

  deploy:
    needs: build
    permissions:
      pages: write     # to deploy to Pages
      id-token: write  # to verify the deployment originates from an appropriate source
    environment:
      name: github-pages
      url: ${{ steps.deployment.outputs.page_url }}
    runs-on: ubuntu-latest
    steps:
      - name: Deploy to GitHub Pages
        id: deployment
        uses: actions/deploy-pages@v2
```

File: .github/workflows/doc.yaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+permissions:
+  contents: read
+  pages: write
+  id-token: write
+concurrency:
+  group: "pages"
+  cancel-in-progress: false
```
Comment (jroimartin) [inline .github/workflows/doc.yaml:20]: I would add a comment to explain why this is needed. Check file comment for more details.

File: build.bash [hunk 6]
```diff
@@ -0,0 +1,80 @@
+# Copyright 2024 Adevinta
+
+if [[ -z $LAVA_VERSION ]]; then
+	echo 'error: missing env var LAVA_VERSION' >&2
+	exit 2
+fi
+
+check_status (){
```
Comment (jroimartin) [inline build.bash]:
```suggestion
check_status() {
```

File: build.bash [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+	local status=$1
+	local error_msg=$2
+	if [[ $status -ne 0 ]]; then
+		echo $error_msg
+		exit $status
+	fi
+}
+
+fix_headers (){
```
Comment (jroimartin) [inline build.bash]:
```suggestion
fix_headers() {
```

File: build.bash [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+	local file=$1
+	sed -i "s/^#/##/" ${file}
+	check_status $? "error: could not fix headers in file ${file}"
+
+}
+
+add_main_header(){
```
Comment (jroimartin) [inline build.bash]:
```suggestion
add_main_header() {
```

File: build.bash [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+	local file=$1
+	local header=$2
+	sed -i "1s/^/# ${header}\n/" ${file}
```
Comment (jroimartin) [inline build.bash]: I think this is more elegant.

```suggestion
	sed -i "1i # ${header}" "${file}"
```

File: build.bash [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+	local error_msg=$2
+	if [[ $status -ne 0 ]]; then
+		echo $error_msg
+		exit $status
+	fi
+}
+
+fix_headers (){
+	local file=$1
+	sed -i "s/^#/##/" ${file}
```
Comment (jroimartin) [inline build.bash]:
```suggestion
	sed -i "s/^#/##/" "${file}"
```

File: .github/workflows/doc.yaml [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+      name: github-pages
+      url: ${{ steps.deployment.outputs.page_url }}
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+      - name: Build documentation
+        run: ./build.bash
+        env:
+          LAVA_VERSION: latest
```
Comment (jroimartin) [inline .github/workflows/doc.yaml]: I think we should always build the documentation for the latest version of Lava. At the end of the day we are only providing documentation for the latest release.

File: build.bash [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+	# Make sure the install directory exists.
+	install_dir="/tmp/bin"
+	mkdir -p "${install_dir}"
+
+	# Try to download a Lava release from GitHub.
+	if (curl -LsSf "${url}" | tar -xz -C "${install_dir}" lava) 2> /dev/null; then
+		return 0
+	fi
+}
```
Comment (jroimartin) [inline build.bash]: I would install the latest Lava release.

File: build.bash [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+# Install Lava.
```
Comment (jroimartin) [inline build.bash]: I would remove this comment. The name of the function is self-explanatory.

```suggestion
# Install Lava.
```

File: build.bash [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+	local url
+	if [[ $version == 'latest' ]]; then
+		url=$GH_DOWNLOAD_URL_LATEST
+	else
+		url=$(printf "${GH_DOWNLOAD_URL_FMT}" "${version}")
+	fi
+
+	# Make sure the install directory exists.
+	install_dir="/tmp/bin"
+	mkdir -p "${install_dir}"
```
Comment (jroimartin) [inline build.bash]: I would move this out of this function and pass the install dir as a parameter of the function before calling it. Also I would generate a directory using `mktemp`.

File: build.bash [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+	topic=$(_jq '.topic')
+	file=$(_jq '.file')
+	header=$(_jq '.header')
+
+	dir=$(dirname ${file})
+
+	mkdir -p ${dir}
+  ${install_dir}/lava help ${topic} > ${file}
+  fix_headers ${file}
+  add_main_header ${file} ${header}
```
Comment (jroimartin) [inline build.bash]: Fix indentation. Use tabs.

File: build.bash [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+done
+
+curl -sSL https://github.com/rust-lang/mdBook/releases/download/v0.4.37/mdbook-v0.4.37-x86_64-unknown-linux-gnu.tar.gz | tar -xz --directory=${install_dir}
+check_status $? "error: could not install mdBook"
+
+sudo chmod +x ${install_dir}/mdbook
```
Comment (jroimartin) [inline build.bash]: Move to a function `install_mdbook` similar to `install_lava`. Also, move the mdbook version to an env var at the beginning of the script. Finally, I would download it with:

```
curl -LsSf "https://github.com/rust-lang/mdBook/releases/download/${MDBOOK_VERSION}/mdbook-${MDBOOK_VERSION}-x86_64-unknown-linux-gnu.tar.gz" | tar -xz -C "${install_dir}" mdbook
```

File: build.bash [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+	_jq() {
+		echo ${row} | jq --raw-output ${1}
+	}
+	topic=$(_jq '.topic')
+	file=$(_jq '.file')
+	header=$(_jq '.header')
+
+	dir=$(dirname ${file})
+
+	mkdir -p ${dir}
```
Comment (jroimartin) [inline build.bash]:
```suggestion
	mkdir -p "$(dirname "${file}")"
```

File: build.bash (hunk 15 above)
Comment (jroimartin) [inline build.bash]:
```suggestion
  ${install_dir}/lava help "${topic}" > "${file}"
  fix_headers "${file}"
  add_main_header "${file}" "${header}"
```

File: src/README.md [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+
+To contribute, please read the [contribution
+guidelines][contributing].
+
+
+[vulcan-docs]: https://adevinta.github.io/vulcan-docs
+[vulcan-checks]: https://github.com/adevinta/vulcan-checks
+[lava-action]: https://github.com/adevinta/lava-action
+[contributing]: https://github.com/adevinta/lava/blob/main/CONTRIBUTING.md
+
```
Comment (jroimartin) [inline src/README.md]: Remove empty line.

**Review by jroimartin — CHANGES_REQUESTED** [review]

**Comment by jroimartin** [comment]
BTW can we include the lava version used to generate the documentation in the site?

Also I would add the LICENSE and the DISCLAIMER files in this repo.

File: .github/workflows/doc.yaml (hunk 4 above)
Comment (seilagamo) [reply .github/workflows/doc.yaml]: This is included in the last commit which means to be dropped before the merging. It's only to test the changes in this branch.

File: build.bash [hunk 19]
```diff
@@ -14,45 +14,40 @@ check_status (){
...
-	mkdir -p "${install_dir}"
+	local dir=$1
+	local url='https://github.com/adevinta/lava/releases/latest/download/lava_linux_amd64.tar.gz'
 
 	# Try to download a Lava release from GitHub.
-	if (curl -LsSf "${url}" | tar -xz -C "${install_dir}" lava) 2> /dev/null; then
+	if (curl -LsSf "${url}" | tar -xz -C "${dir}" lava) 2> /dev/null; then
 		return 0
 	fi
+	sudo chmod +x ${dir}/lava
```
Comment (jroimartin) [inline build.bash]: I think you don't need sudo.

```suggestion
	chmod +x ${dir}/lava
```

File: build.bash [hunk 20]
```diff
@@ -14,45 +14,40 @@ check_status (){
...
+}
+
+install_mdBook() {
+	local dir=$1
+	curl -LsSf "https://github.com/rust-lang/mdBook/releases/download/${MDBOOK_VERSION}/mdbook-${MDBOOK_VERSION}-x86_64-unknown-linux-gnu.tar.gz" | tar -xz -C "${dir}" mdbook
+	check_status $? "error: could not install mdBook"
+	sudo chmod +x ${dir}/mdbook
```
Comment (jroimartin) [inline build.bash]: I think you don't need sudo.

```suggestion
	chmod +x ${dir}/mdbook
```

File: build.bash [hunk 21]
```diff
@@ -14,45 +14,40 @@ check_status (){
...
-		url=$(printf "${GH_DOWNLOAD_URL_FMT}" "${version}")
-	fi
-
-	# Make sure the install directory exists.
-	install_dir="/tmp/bin"
-	mkdir -p "${install_dir}"
+	local dir=$1
+	local url='https://github.com/adevinta/lava/releases/latest/download/lava_linux_amd64.tar.gz'
 
 	# Try to download a Lava release from GitHub.
```
Comment (jroimartin) [inline build.bash:34]: I would remove this comment. Now there is only one download option.

```suggestion
```

File: build.bash [hunk 22]
```diff
@@ -14,45 +14,40 @@ check_status (){
...
+	if (curl -LsSf "${url}" | tar -xz -C "${dir}" lava) 2> /dev/null; then
 		return 0
 	fi
+	sudo chmod +x ${dir}/lava
+}
+
+install_mdBook() {
+	local dir=$1
+	curl -LsSf "https://github.com/rust-lang/mdBook/releases/download/${MDBOOK_VERSION}/mdbook-${MDBOOK_VERSION}-x86_64-unknown-linux-gnu.tar.gz" | tar -xz -C "${dir}" mdbook
+	check_status $? "error: could not install mdBook"
```
Comment (jroimartin) [inline build.bash]: I would follow the same style used in `install_lava`. Also I'd rename the function to `install_mdbook` in lowercase.

```suggestion
        local dir=$1
        local url="https://github.com/rust-lang/mdBook/releases/download/${MDBOOK_VERSION}/mdbook-${MDBOOK_VERSION}-x86_64-unknown-linux-gnu.tar.gz"
        curl -LsSf "${url}" | tar -xz -C "${dir}" mdbook) 2> /dev/null
```

File: build.bash [hunk 23]
```diff
@@ -65,16 +60,18 @@ for row in $(cat pages.json | jq --compact-output '.[]'); do
...
 done
 
-curl -sSL https://github.com/rust-lang/mdBook/releases/download/v0.4.37/mdbook-v0.4.37-x86_64-unknown-linux-gnu.tar.gz | tar -xz --directory=${install_dir}
-check_status $? "error: could not install mdBook"
+${install_dir}/lava version > "src/last_version.md"
+
+if ! install_mdBook "${install_dir}"; then
+	echo 'error: could not install mdBook' >&2
+	exit 1
+fi
```
Comment (jroimartin) [inline build.bash]: I would install mdbook just after installing lava. So both calls to `install_` are together and we fail soon if any of the dependencies is not available. Also, after adding `set -e` you don't need to check the error.

```suggestion
install_mdBook "${install_dir}"
```

File: build.bash [hunk 24]
```diff
@@ -65,16 +60,18 @@ for row in $(cat pages.json | jq --compact-output '.[]'); do
...
+	mkdir -p "$(dirname "${file}")"
+	${install_dir}/lava help "${topic}" > "${file}"
+	check_status $? "error: could not retrieve topic help for topic ${topic}"
+	fix_headers "${file}"
+	add_main_header "${file}" "${header}"
 done
 
-curl -sSL https://github.com/rust-lang/mdBook/releases/download/v0.4.37/mdbook-v0.4.37-x86_64-unknown-linux-gnu.tar.gz | tar -xz --directory=${install_dir}
-check_status $? "error: could not install mdBook"
+${install_dir}/lava version > "src/last_version.md"
```
Comment (jroimartin) [inline build.bash]: As we discussed, I'd try to move this info to the sidebar.

**Review by jroimartin — CHANGES_REQUESTED** [review]

File: build.bash [hunk 25]
```diff
@@ -0,0 +1,77 @@
+# Copyright 2024 Adevinta
+
```
Comment (jroimartin) [inline build.bash:2]: I would add `set -e -u` and remove all calls to `check_status`.

```suggestion

set -e -u

```

File: build.bash [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+install_lava() {
+	local dir=$1
+	local url='https://github.com/adevinta/lava/releases/latest/download/lava_linux_amd64.tar.gz'
+
+	# Try to download a Lava release from GitHub.
+	if (curl -LsSf "${url}" | tar -xz -C "${dir}" lava) 2> /dev/null; then
+		return 0
+	fi
```
Comment (jroimartin) [inline build.bash]: I would remove the comment and the if. `set -e` will capture the error.

```suggestion
	local dir=$1
	local url='https://github.com/adevinta/lava/releases/latest/download/lava_linux_amd64.tar.gz'
	curl -LsSf "${url}" | tar -xz -C "${dir}" lava) 2> /dev/null
```

File: build.bash [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+	check_status $? "error: could not install mdBook"
+	sudo chmod +x ${dir}/mdbook
+}
+
+install_dir=$(mktemp -d)
+
+if ! install_lava "${install_dir}"; then
+	echo 'error: could not install lava' >&2
+	exit 1
+fi
```
Comment (jroimartin) [inline build.bash]: After adding `set -e` you can just call `install_lava`.

```suggestion
install_lava "${install_dir}
```

**Review by jroimartin — CHANGES_REQUESTED** [review]

**Review by jroimartin — CHANGES_REQUESTED** [review]
LGTM. Clean up the commit history.

**Review by jroimartin — APPROVED** [review]

**Pull request merged**